src/include/server/mir/server_configuration.h

```diff
@@ -113,9 +113,25 @@ class ServerConfiguration
     virtual std::shared_ptr<scene::SessionLock> the_session_lock() = 0;
 
     using WaylandProtocolExtensionFilter = std::function<bool(std::shared_ptr<scene::Session> const&, char const*)>;
+
+    [[deprecated("Prefer using set_wayland_extension_policy() to set a policy on a per-interface bassis.")]]
     virtual void set_wayland_extension_filter(WaylandProtocolExtensionFilter const& extension_filter) = 0;
+
+    [[deprecated("Prefer using set_wayland_extension_policy() to set a policy on a per-interface bassis.")]]
     virtual void set_enabled_wayland_extensions(std::vector<std::string> const& extensions) = 0;
 
+    /// Set a policy for the provided wayland extension.
+    ///
+    /// Only clients whose \p policy returns `true` will be allowed to use the interface
+    /// given by \p interface_name.
+    ///
+    /// \param interface_name the interface to apply the policy on
+    /// \param policy the policy predicate with the session and interface name provided
+    ///               as arguments.
+    virtual void set_wayland_extension_policy(
+        std::string const& interface_name,
+        std::function<bool(std::shared_ptr<scene::Session> const&, const char*)> const& policy) = 0;
```

Comment:
This is an ABI break... we need to bump the so number and rearrange the symbols.map